### PR TITLE
ipam/crd: Fix ENI leak due to miscounting of empty interface slots

### LINKED ISSUE #21747: alibabacloud: ENI leak when IP addresses run out in a subnet
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

In a cilium cluster with `ipam=alibabacloud`, we have seen ENI leakage with the following operator log:

```
level=info msg="Attached ENI to instance" eniID=eni-xxx instanceID=i-xxx name=nodexxx securityGroupIDs="[sg-xxx]" subsys=ipam toAllocate=2 vSwitchID=vsw-xxx
level=info msg="Synchronized ENI information" numInstances=348 numSecurityGroups=7 numVPCs=7 numVSwitches=30 subsys=eni
level=info msg="Synchronized ENI information" numInstances=348 numSecurityGroups=7 numVPCs=7 numVSwitches=30 subsys=eni
level=info msg="Resolving IP deficit of node" available=2 availableForAllocation=0 availableInterfaces=1 instanceID=i-xxx maxIPsToAllocate=2 name=nodexxx neededIPs=2 remainingInterfaces=1 selectedInterface= selectedPoolID= subsys=ipam used=2
level=info msg="No more IPs available, creating new ENI" instanceID=i-xxx name=nodexxx securityGroupIDs="[sg-xxx]" subsys=ipam toAllocate=2 vSwitchID=vsw-xxx

level=info msg="Created new ENI" eniID=eni-xxx instanceID=i-xxx name=nodexxx securityGroupIDs="[sg-xxx]" subsys=ipam toAllocate=2 vSwitchID=vsw-xxx
level=warning msg="Unable to create interface on instance: unable to attach ENI SDK.ServerError\nErrorCode: EniPerInstanceLimitExceeded\nRecommend: https://next.api.aliyun.com/troubleshoot?q=EniPerInstanceLimitExceeded&product=Ecs\nRequestId: xxx\nMessage: The number of ENI exceeds the limit(2) for the type of instance you are trying to launch." instanceID=i-xxx name=nodexxx subsys=ipam
level=info msg="Created new ENI" eniID=eni-xxx instanceID=i-xxx name=nodexxx securityGroupIDs="[sg-xxx]" subsys=ipam toAllocate=2 vSwitchID=vsw-xxx
level=warning msg="Unable to create interface on instance: unable to attach ENI SDK.ServerError\nErrorCode: EniPerInstanceLimitExceeded\nRecommend: https://next.api.aliyun.com/troubleshoot?q=EniPerInstanceLimitExceeded&product=Ecs\nRequestId: xxx\nMessage: The number of ENI exceeds the limit(2) for the type of instance you are trying to launch." instanceID=i-xxx name=nodexxx subsys=ipam
```
the following log repeats until all IP addresses in all subnets are consumed, and ENIs created here are leaked in `Available` status (not attached to any instance):
```
level=info msg="Created new ENI" eniID=eni-xxx instanceID=i-xxx name=nodexxx securityGroupIDs="[sg-xxx]" subsys=ipam toAllocate=2 vSwitchID=vsw-xxx
level=warning msg="Unable to create interface on instance: unable to attach ENI SDK.ServerError\nErrorCode: EniPerInstanceLimitExceeded\nRecommend: https://next.api.aliyun.com/troubleshoot?q=EniPerInstanceLimitExceeded&product=Ecs\nRequestId: xxx\nMessage: The number of ENI exceeds the limit(2) for the type of instance you are trying to launch." instanceID=i-xxx name=nodexxx subsys=ipam
```

### Cilium Version

v1.10.7

### Kernel Version

5.10.56

### Kubernetes Version

v1.19.3

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jaffcheng**
please see commit msg

Fixes: #21747

```release-note
Fix ENI leak in Alibaba due to miscounting of empty interface slots
```

### CODE REVIEW

File: pkg/ipam/metrics/metrics.go [hunk 1]
```diff
@@ -14,14 +14,17 @@ import (
...
-	ReleaseIpOps          *prometheus.CounterVec
-	IPsAllocated          *prometheus.GaugeVec
+	registry             *prometheus.Registry
+	Allocation           *prometheus.HistogramVec
+	Release              *prometheus.HistogramVec
+	AllocateInterfaceOps *prometheus.CounterVec
+	AllocateIpOps        *prometheus.CounterVec
+	ReleaseIpOps         *prometheus.CounterVec
+	IPsAllocated         *prometheus.GaugeVec
+	// Deprecated: Use InterfaceCandidates and EmptyInterfaceSlots instead
```
Comment (qmonnet) [inline pkg/ipam/metrics/metrics.go]: Should we plan to remove it in the next version?

```suggestion
	// Deprecated, will be removed in version 1.14:
	// Use InterfaceCandidates and EmptyInterfaceSlots instead
```

**Review by qmonnet — APPROVED** [review]
Looks good as far as I can tell, but I'm not familiar with Alibaba's IPAM.

From the issue report I understand the bug has been present since at least 1.10, is this fix something that we should consider for backports?

**Comment by jaffcheng** [comment]
Thanks for the comment.
> From the issue report I understand the bug has been present since at least 1.10, is this fix something that we should consider for backports?

Yes, I think this needs to be backported. I'm a little surprised this issue hasn't been reported by another user, since it should affect AWS ENI as well if I understand correctly.

**Review by christarazi — APPROVED** [review]
Fix makes sense to me, thanks! Should we go a step further and ensure that we deallocate the ENI if we fail to attach it?

**Review by tommyp1ckles — APPROVED** [review]
Azure changes LGTM

**Comment by jaffcheng** [comment]
> Should we go a step further and ensure that we deallocate the ENI if we fail to attach it?

@christarazi Thanks for the advice! Just realized we already have ENI rollback logic, but alibabacloud missed it after a failed `AttachNetworkInterface` API call. This could explain why AWS users haven't run into this.
Added a commit to fix this, please take another look.

**Comment by christarazi** [comment]
Looks good, thanks!

**Comment by pchaigno** [comment]
Removing backport labels here as this PR has upgrade impact and deprecations. Please yell if some part of the PR does need to be bsckported.

**Comment by christarazi** [comment]
@jaffcheng Do we need to do the same thing for AWS ENI?

**Comment by jaffcheng** [comment]
I think this fix should also affect AWS ENI. IIUC, if an AWS node encounters this, the cilium-operator without this PR would keep creating and attaching(limit error) and deleting an ENI, but no ENI leak would happen.

**Comment by christarazi** [comment]
@jaffcheng Could you open a PR to resolve discrepancies between AWS and Alibaba?

**Comment by jaffcheng** [comment]
> @jaffcheng Could you open a PR to resolve discrepancies between AWS and Alibaba?

@christarazi Sorry for my wording not being clear enough, I think this PR should have fixed the ENI slot miscounting problem for AWS:  

https://github.com/cilium/cilium/commit/118c14819f3bfb2e6ec59f0883e5873b81a22691#diff-30611d880237aa1610d4ba922a8639e304d4f1e398b98418bdd4200ee4536066R225-R252

The discrepancies between Alibaba and AWS are fixed in this commit:  
https://github.com/cilium/cilium/pull/21800/commits/8e23fe4f44d342e7ad64bc547184f581da872bb2

**Comment by christarazi** [comment]
Ah perfect, thanks @jaffcheng!

**Pull request merged**